File: internal/store/store.go
package store

import (
	"encoding/json"
	"os"
	"sync"
	"teleport-ui/internal/models"
)

// DataStore abstracts the file reading and writing logic.
type DataStore struct {
	mu         sync.RWMutex
	dbPath     string
	configPath string
}

func NewDataStore(dbPath, configPath string) *DataStore {
	// Initialize default databases.json if not exist
	if _, err := os.Stat(dbPath); os.IsNotExist(err) {
		_ = os.WriteFile(dbPath, []byte("[]\n"), 0644)
	}

	// Initialize default config.json if not exist
	if _, err := os.Stat(configPath); os.IsNotExist(err) {
		cfg := models.GlobalConfig{TeleportProxy: ""}
		out, _ := json.MarshalIndent(cfg, "", "  ")
		_ = os.WriteFile(configPath, out, 0644)
	}

	return &DataStore{
		dbPath:     dbPath,
		configPath: configPath,
	}
}

func (s *DataStore) GetDatabases() ([]models.Database, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	b, err := os.ReadFile(s.dbPath)
	if err != nil {
		return []models.Database{}, nil // Fallback safe
	}

	if len(b) == 0 {
		return []models.Database{}, nil
	}

	dbs := []models.Database{} // Strongly typed empty allocation guarantees [] instead of null in JSON
	if err := json.Unmarshal(b, &dbs); err != nil {
		return []models.Database{}, nil // Treat as empty if corrupted
	}
	return dbs, nil
}

func (s *DataStore) SaveDatabase(db models.Database) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	b, _ := os.ReadFile(s.dbPath)
	var dbs []models.Database
	if len(b) > 0 {
		_ = json.Unmarshal(b, &dbs)
	}

	dbs = append(dbs, db)

	out, err := json.MarshalIndent(dbs, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(s.dbPath, out, 0644)
}

func (s *DataStore) DeleteDatabase(id string) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	b, _ := os.ReadFile(s.dbPath)
	var dbs []models.Database
	if len(b) > 0 {
		_ = json.Unmarshal(b, &dbs)
	}

	updated := []models.Database{}
	for _, db := range dbs {
		if db.ID != id {
			updated = append(updated, db)
		}
	}

	out, err := json.MarshalIndent(updated, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(s.dbPath, out, 0644)
}

func (s *DataStore) GetConfig() (models.GlobalConfig, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	var cfg models.GlobalConfig
	b, err := os.ReadFile(s.configPath)
	if err != nil || len(b) == 0 {
		return models.GlobalConfig{TeleportProxy: ""}, nil
	}

	if err := json.Unmarshal(b, &cfg); err != nil {
		return models.GlobalConfig{TeleportProxy: ""}, nil
	}

	return cfg, nil
}

func (s *DataStore) SaveConfig(cfg models.GlobalConfig) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	out, err := json.MarshalIndent(cfg, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(s.configPath, out, 0644)
}

File: src-tauri/src/store.rs
use crate::models::{Database, GlobalConfig};
use serde_json;
use std::fs;
use std::path::PathBuf;
use std::sync::Mutex;

pub struct DataStore {
    mu: Mutex<()>,
    db_path: PathBuf,
    config_path: PathBuf,
}

impl DataStore {
    pub fn new(db_path: PathBuf, config_path: PathBuf) -> Self {
        // Ensure parent dirs exist
        if let Some(parent) = db_path.parent() {
            let _ = fs::create_dir_all(parent);
        }
        if let Some(parent) = config_path.parent() {
            let _ = fs::create_dir_all(parent);
        }

        // Initialize databases.json if not exists
        if !db_path.exists() {
            let _ = fs::write(&db_path, b"[]\n");
        }

        // Initialize config.json if not exists
        if !config_path.exists() {
            let cfg = GlobalConfig::default();
            if let Ok(json) = serde_json::to_string_pretty(&cfg) {
                let _ = fs::write(&config_path, json);
            }
        }

        Self {
            mu: Mutex::new(()),
            db_path,
            config_path,
        }
    }

    pub fn get_databases(&self) -> Vec<Database> {
        let _lock = self.mu.lock().unwrap();
        let bytes = fs::read(&self.db_path).unwrap_or_default();
        if bytes.is_empty() {
            return vec![];
        }
        serde_json::from_slice(&bytes).unwrap_or_default()
    }

    pub fn save_database(&self, db: Database) -> Result<(), String> {
        let _lock = self.mu.lock().unwrap();
        let bytes = fs::read(&self.db_path).unwrap_or_default();
        let mut dbs: Vec<Database> = if bytes.is_empty() {
            vec![]
        } else {
            serde_json::from_slice(&bytes).unwrap_or_default()
        };

        // Replace if id exists, else append
        let pos = dbs.iter().position(|d| d.id == db.id);
        if let Some(i) = pos {
            dbs[i] = db;
        } else {
            dbs.push(db);
        }

        let json = serde_json::to_string_pretty(&dbs).map_err(|e| e.to_string())?;
        fs::write(&self.db_path, json).map_err(|e| e.to_string())
    }

    pub fn delete_database(&self, id: &str) -> Result<(), String> {
        let _lock = self.mu.lock().unwrap();
        let bytes = fs::read(&self.db_path).unwrap_or_default();
        let mut dbs: Vec<Database> = if bytes.is_empty() {
            vec![]
        } else {
            serde_json::from_slice(&bytes).unwrap_or_default()
        };

        dbs.retain(|d| d.id != id);

        let json = serde_json::to_string_pretty(&dbs).map_err(|e| e.to_string())?;
        fs::write(&self.db_path, json).map_err(|e| e.to_string())
    }

    pub fn get_config(&self) -> GlobalConfig {
        let _lock = self.mu.lock().unwrap();
        let bytes = fs::read(&self.config_path).unwrap_or_default();
        if bytes.is_empty() {
            return GlobalConfig::default();
        }
        serde_json::from_slice(&bytes).unwrap_or_default()
    }

    pub fn save_config(&self, cfg: GlobalConfig) -> Result<(), String> {
        let _lock = self.mu.lock().unwrap();
        let json = serde_json::to_string_pretty(&cfg).map_err(|e| e.to_string())?;
        fs::write(&self.config_path, json).map_err(|e| e.to_string())
    }
}
